hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -1200,11 +1204,13 @@ transpose2D = transpose @0 @1
 -- >>> dtype &&& shape $ diag @1 Lower (ones :: CPUTensor 'D.Float '[3,2])
 -- (Float,[2])
 diag
-  :: forall (index :: Nat) (i :: Nat) (j :: Nat) device dtype
-   . KnownNat index
+  :: forall (index :: Nat) (shape :: [Nat]) (shape' :: [Nat]) device dtype
+   . ( KnownNat index
+     , shape' ~ DiagShape index shape
+     )
   => Tri -- ^ upper or lower
-  -> Tensor device dtype '[i, j] -- ^ input
-  -> Tensor device dtype '[Min i j - index] -- ^ output
+  -> Tensor device dtype shape -- ^ input
+  -> Tensor device dtype shape' -- ^ output
 diag tri t = unsafePerformIO $ ATen.cast2 ATen.Managed.tensor_diag_l t
```

Comment:
you may want to look into writing a test suite for this function in `Torch.Typed.FunctionalSpec`, see, e.g., https://github.com/hasktorch/hasktorch/pull/409/files#diff-e428518635181a3dbe949a37402aa368

Reply:
for instance, the type signature claims that this function works for all `device`s and all `dtype`s. you may want to make sure that this is accurate :)

Reply:
on it, thanks!